### PR TITLE
[aggregator] Exchange on read

### PR DESCRIPTION
**Pull request by georgemitenkov**
This PR adds support for value exchange and ID generation
in LatestStateView. Right now we only exchange `u128`, but
we can support more types with #10128.

### CODE REVIEW

File: aptos-move/block-executor/src/view.rs [hunk 1]
```diff
@@ -353,7 +425,48 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TAggregator
         state_key: &Self::IdentifierV1,
         _mode: AggregatorReadMode,
     ) -> anyhow::Result<Option<StateValue>> {
-        // TODO: Integrate aggregators.
+        // TODO: Integrate aggregators V1.
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/view.rs:428]: I don't understand this TODO from before, nor the updated one. :) What is exactly needed here? doesn't this work correctly for V1?

File: aptos-move/block-executor/src/view.rs [hunk 2]
```diff
@@ -353,7 +425,48 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TAggregator
...
+        &self,
+        layout: &MoveTypeLayout,
+        value_to_exchange: Value,
+    ) -> TransformationResult<Value> {
+        let identifier = self.generate_aggregator_v2_id();
+        match &self.latest_view {
+            ViewState::Sync(state) => {
+                state.set_aggregator_v2_lifted_value(identifier.clone(), value_to_exchange)
+            },
+            ViewState::Unsync(_state) => {},
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/view.rs:462]: Add a TODO for unsync, I'll be adding aggregator maps there as well

File: aptos-move/block-executor/src/view.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:428]: We can lift aggregator value using MoveTypeLayout::U128, so we redirect to V2 flow, and the V1 state item is treated as data.

File: aptos-move/block-executor/src/view.rs [hunk 3]
```diff
@@ -68,11 +79,21 @@ impl<'a, T: Transaction, X: Executable> ParallelState<'a, T, X> {
...
             versioned_map: shared_map,
             scheduler: shared_scheduler,
-            _counter: shared_counter,
+            counter: shared_counter,
             captured_reads: RefCell::new(Vec::new()),
         }
     }
 
+    fn set_aggregator_v2_lifted_value(&self, id: T::Identifier, value: Value) {
+        // TODO: Integrate snapshots.
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:88]: I ll rebase on your PR then

File: aptos-move/block-executor/src/view.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:428]: Made TODO more clear

File: aptos-move/aptos-aggregator/src/delta_change_set.rs [hunk 4]
```diff
@@ -416,6 +416,10 @@ mod test {
                 Some("Error message from BadStorage.".to_string()),
             )))
         }
+
+        fn generate_aggregator_v2_id(&self) -> Self::IdentifierV2 {
```
Comment (vusirikala) [inline aptos-move/aptos-aggregator/src/delta_change_set.rs:420]: `delta_change_set` is more of an aggregator v1 file. Can we have this function in a aggregator v2 file?

File: aptos-move/aptos-aggregator/src/resolver.rs [hunk 5]
```diff
@@ -82,9 +82,7 @@ pub trait TAggregatorView {
 
     /// Returns a unique per-block identifier that can be used when creating a
     /// new aggregator V2.
-    fn generate_aggregator_v2_id(&self) -> Self::IdentifierV2 {
-        unimplemented!("ID generation for Aggregator V2 is not yet supported")
-    }
+    fn generate_aggregator_v2_id(&self) -> Self::IdentifierV2;
```
Comment (vusirikala) [inline aptos-move/aptos-aggregator/src/resolver.rs:85]: Just trying to learn. What's the difference between this and using "unimplemented" as before?

File: aptos-move/aptos-aggregator/src/tests/types.rs [hunk 6]
```diff
@@ -17,11 +17,21 @@ pub fn aggregator_v1_state_key_for_test(key: u128) -> StateKey {
     StateKey::raw(key.to_le_bytes().to_vec())
 }
 
-#[derive(Default)]
 pub struct FakeAggregatorView {
     // TODO: consider adding deltas to test different read modes.
     v1_store: HashMap<StateKey, StateValue>,
     v2_store: HashMap<AggregatorID, u128>,
+    counter: RefCell<u32>,
```
Comment (vusirikala) [inline aptos-move/aptos-aggregator/src/tests/types.rs:24]: Why RefCell<u32> instead of u32?

File: aptos-move/aptos-aggregator/src/types.rs [hunk 7]
```diff
@@ -26,6 +26,20 @@ impl AggregatorID {
...
+// Used for ID generation and for Value exchange.
+impl From<u64> for AggregatorID {
+    fn from(value: u64) -> Self {
+        Self::new(value)
+    }
+}
+
+// Used for Value exchange.
+impl From<AggregatorID> for u64 {
+    fn from(id: AggregatorID) -> Self {
```
Comment (vusirikala) [inline aptos-move/aptos-aggregator/src/types.rs:38]: Should be returning u64.

Comment (georgemitenkov) [reply aptos-move/aptos-aggregator/src/types.rs:38]: It compiled though, Rust magic I guess 😂

File: aptos-move/aptos-aggregator/src/tests/types.rs (hunk 6 above)
Comment (georgemitenkov) [reply aptos-move/aptos-aggregator/src/tests/types.rs:24]: RefCell allows interior mutability, while otherwise you would need &mut

File: aptos-move/aptos-aggregator/src/resolver.rs (hunk 5 above)
Comment (georgemitenkov) [reply aptos-move/aptos-aggregator/src/resolver.rs:85]: Before we had a default implementation (or better to say no implementation), now we require one

File: aptos-move/block-executor/src/executor.rs [hunk 8]
```diff
@@ -644,8 +646,8 @@ where
         let executor = E::init(executor_arguments);
         drop(init_timer);
 
+        let counter = Rc::new(RefCell::new(0));
```
Comment (vusirikala) [inline aptos-move/block-executor/src/executor.rs:649]: I'm actually confused with the counter. I thought there would be one shared counter that will be used by everyone. But there is a counter here in executor.rs, there is another counter in executor_view_base.rs, there is another counter in ParallelState/SequentialState, etc.

File: aptos-move/block-executor/src/executor.rs [hunk 9]
```diff
@@ -655,7 +657,7 @@ where
                 base_view,
                 ViewState::Unsync(SequentialState {
                     unsync_map: &data_map,
-                    _counter: &0,
+                    counter: Rc::clone(&counter),
```
Comment (vusirikala) [inline aptos-move/block-executor/src/executor.rs:660]: If we are using Rc pointer here for SequentialState, shouldn't we be doing the same for ParallelState as well.

File: aptos-move/block-executor/src/view.rs [hunk 10]
```diff
@@ -13,7 +13,9 @@ use aptos_aggregator::{
 };
 use aptos_logger::error;
 use aptos_mvhashmap::{
-    types::{MVDataError, MVDataOutput, MVModulesError, MVModulesOutput, TxnIndex},
+    types::{
+        MVAggregatorsError, MVDataError, MVDataOutput, MVModulesError, MVModulesOutput, TxnIndex,
+    },
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:18]: Generally speaking, is there any advantage in using so many Error types?

File: aptos-move/mvhashmap/src/unit_tests/mod.rs [hunk 11]
```diff
@@ -54,7 +54,8 @@ fn create_write_read_placeholder_struct() {
     let ap2 = KeyType(b"/foo/c".to_vec());
     let ap3 = KeyType(b"/foo/d".to_vec());
 
-    let mvtbl: MVHashMap<KeyType<Vec<u8>>, usize, TestValue, ExecutableTestType> = MVHashMap::new();
+    let mvtbl: MVHashMap<KeyType<Vec<u8>>, usize, usize, TestValue, ExecutableTestType> =
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/unit_tests/mod.rs:57]: Why is identifier type usize here instead of u64?

File: aptos-move/mvhashmap/src/unit_tests/proptest_types.rs [hunk 12]
```diff
@@ -211,7 +211,7 @@ where
 
     let baseline = Baseline::new(transactions.as_slice(), test_group);
     // Only testing data, provide executable type ().
-    let map = MVHashMap::<KeyType<K>, usize, Value<V>, ExecutableTestType>::new();
+    let map = MVHashMap::<KeyType<K>, usize, usize, Value<V>, ExecutableTestType>::new();
```
Comment (vusirikala) [inline aptos-move/mvhashmap/src/unit_tests/proptest_types.rs:214]: Why is identifier type usize here instead of u64?

File: aptos-move/block-executor/src/view.rs [hunk 13]
```diff
@@ -229,6 +257,30 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
...
+    /// Given a state value, performs deserialization-serialization round-trip
+    /// to replace any aggregator / aggregator snapshot values.
+    fn lift_aggregators_v2_from_state_value(
+        &self,
+        state_value: StateValue,
+        layout: &MoveTypeLayout,
+    ) -> anyhow::Result<StateValue> {
+        state_value.map_bytes(|bytes| {
+            let patched_value =
+                deserialize_and_exchange(bytes.as_ref(), layout, self).ok_or_else(|| {
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:270]: The word exchange is a little bit confusing. It could either id --> value or value --> id. It would be nice to use a more explicit name.

File: aptos-move/block-executor/src/view.rs [hunk 14]
```diff
@@ -229,6 +257,30 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
         }
         ret
     }
+
+    /// Given a state value, performs deserialization-serialization round-trip
+    /// to replace any aggregator / aggregator snapshot values.
+    fn lift_aggregators_v2_from_state_value(
+        &self,
+        state_value: StateValue,
+        layout: &MoveTypeLayout,
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:266]: I'm not sure. Should this be Option<MoveTypeLayout>?

File: aptos-move/block-executor/src/view.rs [hunk 15]
```diff
@@ -240,7 +292,7 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
     fn get_resource_state_value(
         &self,
         state_key: &Self::Key,
-        _maybe_layout: Option<&Self::Layout>,
+        maybe_layout: Option<&Self::Layout>,
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:295]: We used MoveTypeLayout in other functions in the file. Is Self::Layout different than MoveTypeLayout?

File: aptos-move/block-executor/src/view.rs [hunk 16]
```diff
@@ -254,12 +306,39 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
...
+                        (from_storage, _) => from_storage,
+                    };
 
                     // This base value can also be used to resolve AggregatorV1 directly from
                     // the versioned data-structure (without more storage calls).
                     state.versioned_map.data().provide_base_value(
                         state_key.clone(),
-                        TransactionWrite::from_state_value(from_storage),
+                        TransactionWrite::from_state_value(maybe_patched_from_storage),
                     );
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:342]: Didn't understand this statement. We are creating a transaction write when a resource is read?

File: aptos-move/block-executor/src/view.rs [hunk 17]
```diff
@@ -353,7 +432,82 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TAggregator
...
+    }
+}
+
+// For aggregators V2, values are replaced with identifiers at deserialization time,
+// and are replaced back when the value is serialized. The "lifted" values are cached
+// by the `LatestView` in the aggregators multi-version data structure.
+impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> ValueExchange
+    for LatestView<'a, T, S, X>
+{
+    fn try_exchange(
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:461]: Same here. Exchange could mean id --> value and value --> id. It would be nice to use a more explicit name.

File: aptos-move/block-executor/src/view.rs [hunk 18]
```diff
@@ -353,7 +432,82 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TAggregator
...
+        &self,
+        layout: &MoveTypeLayout,
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:463]: Should this be Option<MoveTypeLayout>?

File: aptos-move/block-executor/src/view.rs [hunk 19]
```diff
@@ -353,7 +432,82 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TAggregator
...
+        value_to_exchange: Value,
+    ) -> TransformationResult<Value> {
+        let identifier = self.generate_aggregator_v2_id();
+        match &self.latest_view {
+            ViewState::Sync(state) => {
+                // TODO(aggregator): Integrate other types as well.
+                let base_value = value_to_exchange
```
Comment (vusirikala) [inline aptos-move/block-executor/src/view.rs:470]: This was my understanding from the name "try_exchange". I thought `value_to_exchange` could be any resource, and this resource could contain many aggregators. We need to exchange each of the aggregators in the resource. 
But it seems like `value_to_exchange` could only be a `u128`?

File: aptos-move/aptos-aggregator/src/types.rs (hunk 7 above)
Comment (igor-aptos) [reply aptos-move/aptos-aggregator/src/types.rs:38]: well Self is u64 - because this is implementation of function for u64 type :)

File: aptos-move/block-executor/src/executor.rs (hunk 8 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:649]: Just to make sure I understood correctly - this counter is for when we are doing sequential execution, and the other one is for when we are doing parallel execution? 
Is ExecutorViewBase not used in sequential execution?


and no matter which execution is used, there is only a single counter in existance, correct?

File: aptos-move/block-executor/src/executor.rs (hunk 9 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:660]: sequential doesn't need to deal with threads, so doesn't need to be Atomic. Basically Rc<RefCell<>> is a looser non-thread safe variant, and Atomic is a stronger thread safe variant.

We could use Atomic here (with unnecessary overhead). we cannot use Rc<RefCell<>> in parallel execution

File: aptos-move/block-executor/src/task.rs [hunk 20]
```diff
@@ -47,7 +47,19 @@ pub trait Transaction: Sync + Send + Clone + 'static {
...
+        + Sync
+        + Copy
+        + Clone
+        + Hash
+        + Eq
+        + Debug
+        // Identifiers are generated from u32 and values are exchanged with u64.
+        + From<u64>
+        // Value exchange uses u64.
+        + Into<u64>;
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/task.rs:62]: hm, if you are using this, how are you going to exchange AggregatorSnpshot<String> ?

File: aptos-move/block-executor/src/view.rs (hunk 19 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/view.rs:470]: my understanding is that this is like a visitor pattern - try_exchange is called on every instance of aggregator - i.e. if resource has 3 aggregators, this function will be called 3 times.

File: aptos-move/block-executor/src/view.rs (hunk 10 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:18]: These errors come from different mvhashmaps, and having them with different types seems cleaner in this case.

File: aptos-move/mvhashmap/src/unit_tests/mod.rs (hunk 11 above)
Comment (georgemitenkov) [reply aptos-move/mvhashmap/src/unit_tests/mod.rs:57]: No reason, it doesn't matter to much for unit tests, used usize like it was done for tags, can switch to u64 to make it clearer

File: aptos-move/mvhashmap/src/unit_tests/proptest_types.rs (hunk 12 above)
Comment (georgemitenkov) [reply aptos-move/mvhashmap/src/unit_tests/proptest_types.rs:214]: Same as above

File: aptos-move/block-executor/src/view.rs (hunk 13 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:270]: Maybe `deserialize_and_replace_with_id`?

File: aptos-move/block-executor/src/view.rs (hunk 14 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:266]: This is only called if we have Some(layout), so having concrete reference here is fine

File: aptos-move/block-executor/src/executor.rs (hunk 8 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/executor.rs:649]: Yes, so we need to support
1. Parallel execution
2. Sequential execution
3. Simulating txns, e.g. running view functions, etc. or tests without block-stm.

For (1) we use an atomic which is passed to ParallelState, for (2) we use this and pass to `Unsync` map, for (3) I added a counter to the adapter which makes the state view be treated as executor view, to be able to simulate txns, etc. So it is always 1 counter in existence.

File: aptos-move/block-executor/src/view.rs (hunk 15 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:295]: Self::Layout is an associated type for the trait (https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html), so it is actually the same as MoveTypeLayout for this implementation

File: aptos-move/block-executor/src/view.rs (hunk 16 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:342]: This is not a write, `provide_base_value` caches the value from storage, and stores it at artificial 0 index (or -1 sort of). But it is not registered as a "write".

File: aptos-move/block-executor/src/view.rs (hunk 17 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:461]: The name is generic because you can do arbitrary value to value replacement. It's just as using it for IDs. I guess there are no other usages anyway, so renaming might be indeed better👍

Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:461]: Actually, looking at that, this is pretty generic, haha, so I think exchange is better. Let me know if you really prefer values/ids

File: aptos-move/block-executor/src/view.rs (hunk 18 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:463]: It's a trait method, and a trait simply allows you to replace a value with some particular layout. So it cannot be option.

File: aptos-move/block-executor/src/view.rs (hunk 19 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:470]: Yes, we have a visitor that does serialization/deserialization, and exchange is just added on top. SO it replaces the current node which is visited.

File: aptos-move/block-executor/src/task.rs (hunk 20 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/task.rs:62]: This is for identifiers? They are always u64, but because we use AggregatorID wrappers, we need to be able to get the inner value.

File: aptos-move/block-executor/src/view.rs (hunk 13 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:270]: If we want to keep exchange generic, we need the old name though. I added a detailed comment there

File: aptos-move/block-executor/src/task.rs (hunk 20 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/task.rs:62]: I see, so you take a identifier, transform into u64, and then do to_string if it is AggregatorSnapshot<String>

File: aptos-move/aptos-aggregator/src/types.rs [hunk 21]
```diff
@@ -23,11 +48,118 @@ impl AggregatorID {
...
+        }
+    };
+
+    let is_vec_u8 =
+        |l: &MoveTypeLayout| -> bool { matches!(l, Vector(b) if matches!(b.as_ref(), U8)) };
+
+    single_field_struct_fields(layout)
+        .map(|ls| is_vec_u8(ls.get(0).expect("A single field must exist")))
+        .unwrap_or(false)
+}
```
Comment (igor-aptos) [inline aptos-move/aptos-aggregator/src/types.rs:83]: maybe this is a bit cleaner?

```
fn is_string_layout<'a>(layout: &'a MoveTypeLayout) -> bool {
    use MoveTypeLayout::*;
    if let Struct(move_struct) = layout {
        if let [Vector(elem)] = move_struct.fields().as_slice() {
            if let U8 = elem.as_ref() {
                return true;
            }
        }
    }
    false
}
```

File: aptos-move/aptos-aggregator/src/types.rs [hunk 22]
```diff
@@ -23,11 +48,118 @@ impl AggregatorID {
...
+                return Err(PartialVMError::new(StatusCode::VALUE_SERIALIZATION_ERROR)
+                    .with_message(format!(
+                        "Failed to convert a Move value with {} layout into an identifier",
+                        layout
+                    )))
+            },
+        }))
+    }
+}
+
```
Comment (igor-aptos) [inline aptos-move/aptos-aggregator/src/types.rs:162]: this all seems similar to the code in aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

maybe worth deduping if possible? (fine if to do in a separate PR)

File: aptos-move/aptos-vm/src/tests/mock_view.rs [hunk 23]
```diff
@@ -70,66 +70,82 @@ impl MockStateView {
...
+        let identifier_value = match layout {
+            MoveTypeLayout::U64 => Value::u64(identifier),
+            MoveTypeLayout::U128 => Value::u128(identifier as u128),
+            _ => {
+                return Err(TransformationError::new(format!(
+                    "Cannot embed identifier for {}",
+                    layout
+                )))
+            },
+        };
```
Comment (igor-aptos) [inline aptos-move/aptos-vm/src/tests/mock_view.rs:108]: should we reuse the utilities? (and so it works with strings too

File: aptos-move/aptos-vm/src/tests/mock_view.rs [hunk 24]
```diff
@@ -70,66 +70,82 @@ impl MockStateView {
...
+        let identifier = match layout {
+            MoveTypeLayout::U64 => identifier.value_as::<u64>().expect(""),
+            MoveTypeLayout::U128 => identifier.value_as::<u128>().expect("") as u64,
+            _ => {
+                return Err(TransformationError::new(format!(
+                    "Value {} cannot be an identifier",
+                    identifier
+                )))
+            },
+        };
```
Comment (igor-aptos) [inline aptos-move/aptos-vm/src/tests/mock_view.rs:129]: same, should we reuse utilities?

**Review by igor-aptos — APPROVED** [review]
added inline comments, otherwise looks good

File: aptos-move/block-executor/src/task.rs [hunk 25]
```diff
@@ -80,11 +78,7 @@ pub trait ExecutorTask: Sync {
     /// Execute a single transaction given the view of the current state.
     fn execute_transaction(
         &self,
-        view: &impl TExecutorView<
-            <Self::Txn as Transaction>::Key,
-            MoveTypeLayout,
-            <Self::Txn as Transaction>::Identifier,
-        >,
+        view: &impl TExecutorView<<Self::Txn as Transaction>::Key, MoveTypeLayout, AggregatorID>,
```
Comment (gelash) [inline aptos-move/block-executor/src/task.rs:81]: I commented on @igor-aptos 's PR that it might be okay to specialize Block executor on AggregatorID, at least for now, but want to make sure we are consistent with the decision on this?

File: aptos-move/block-executor/src/executor.rs (hunk 9 above)
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:660]: Refcell makes sense, missed it 🤦 But why Rc if & would do?

File: aptos-move/block-executor/src/view.rs [hunk 26]
```diff
@@ -254,12 +310,38 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
...
                     let from_storage = self.base_view.get_state_value(state_key)?;
+                    let maybe_patched_from_storage = match (from_storage, maybe_layout) {
+                        // There are aggregators / aggregator snapshots in the
+                        // resource, so we have to replace the actual values with
+                        // identifiers.
+                        // TODO(aggregator): gate by the flag.
+                        (Some(state_value), Some(layout)) => {
+                            let res = self.replace_values_with_identifiers(state_value, layout);
+                            if let Err(err) = &res {
+                                // TODO(aggregator): This means replacement failed
```
Comment (gelash) [inline aptos-move/block-executor/src/view.rs:321]: @igor-aptos has a nice new Error for this returned from Aggregator MVHashMap, we should re-use the flow / types for all of those.

File: aptos-move/block-executor/src/task.rs (hunk 25 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/task.rs]: Specializing is good, I had it before but then removed because of the other PRs.

File: aptos-move/block-executor/src/view.rs (hunk 26 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/view.rs:321]: Do you mean we want to re-use `CommitError` here? Here if we fail (which should not happen), we always have to fallback to recovery mode

File: aptos-move/aptos-aggregator/src/types.rs (hunk 22 above)
Comment (georgemitenkov) [reply aptos-move/aptos-aggregator/src/types.rs]: I made it nicer, and refactored some code into common utils. I still think we can do it better, but keeping for now because the rest can come separately

File: aptos-move/block-executor/src/view.rs [hunk 27]
```diff
@@ -283,9 +311,38 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
...
+                    )
+                })
+                .map(|b| b.into())
+        })
+    }
+
     fn get_resource_state_value_impl(
         &self,
         state_key: &T::Key,
+        maybe_layout: Option<&MoveTypeLayout>,
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/view.rs:345]: @gelash  - with rebase, I needed to add maybe_layout back here that you removed in your PR, does this look correct?

File: aptos-move/block-executor/src/view.rs [hunk 28]
```diff
@@ -366,7 +452,8 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
         &self,
         state_key: &Self::Key,
     ) -> anyhow::Result<Option<StateValueMetadataKind>> {
-        self.get_resource_state_value_impl(state_key, ReadKind::Metadata)
+        // TODO check that passing None here is correct
+        self.get_resource_state_value_impl(state_key, None, ReadKind::Metadata)
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/view.rs:456]: @gelash also these changes - we don't need layout for non-value read kinds, right? (as we don't do exchanges)

**Review by igor-aptos — COMMENTED** [review]
rebased, but CapturedRead changes are a bit conflicting, so asking inline:

File: aptos-move/block-executor/src/view.rs (hunk 28 above)
Comment (gelash) [reply aptos-move/block-executor/src/view.rs:456]: we shouldn't
